In [17]:
from notebook.services.config import ConfigManager

ConfigManager().update('notebook', {'limit_output': 20})

{'limit_output': 20}

In [2]:
from IPython.display import Markdown, display
import time
import pickle

import matplotlib.pyplot as plt
%matplotlib inline

from ipywidgets import interactive, FloatSlider


#import sys
#sys.path.append("../")

from tqdm import tqdm

import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 
#import missingno as msno

import warnings
warnings.filterwarnings('ignore')

In [3]:
# import StandardScaler to transform to mean 0, standard deviation 1
from sklearn.preprocessing import StandardScaler

# import machine learning algorithms
from sklearn.linear_model import LogisticRegression
from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV

from sklearn.metrics import accuracy_score, balanced_accuracy_score, f1_score

## Load Dataframe

In [4]:

df_new = pd.read_csv("german_credit_data_prepared.csv")
#cols_to_drop = ["gender_male", "marital_status_divorced/separated", "marital_status_divorced/separated/married",
#                "marital_status_married/widowed", "marital_status_single"]
#df_new.drop(cols_to_drop, axis=1, inplace=True)
df_new.head()

Duration in month  Credit amount  \
0                  6           1169   
1                 48           5951   
2                 12           2096   
3                 42           7882   
4                 24           4870   

   Installment rate in percentage of disposable income  \
0                                                  4     
1                                                  2     
2                                                  2     
3                                                  2     
4                                                  3     

   Present residence since  Number of existing credits at this bank  \
0                        4                                        2   
1                        2                                        1   
2                        3                                        1   
3                        4                                        1   
4                        4                                        2   

   Number of people being liable to provide maintenance for  gender_male  \
0                                                  1                   1   
1                                                  1                   0   
2                                                  2                   1   
3                                                  2                   1   
4                                                  2                   1   

   marital_status_divorced/separated  \
0                                  0   
1                                  0   
2                                  0   
3                                  0   
4                                  0   

   marital_status_divorced/separated/married  marital_status_married/widowed  \
0                                          0                               0   
1                                          1                               0   
2                                          0                               0   
3                                          0                               0   
4                                          0                               0   

   ...  Housing_rent  Job_management/ highly qualified employee  \
0  ...             0                                          0   
1  ...             0                                          0   
2  ...             0                                          0   
3  ...             0                                          0   
4  ...             0                                          0   

   Job_skilled employee / official  Job_unemployed/ unskilled  - non-resident  \
0                                1                                          0   
1                                1                                          0   
2                                0                                          0   
3                                1                                          0   
4                                1                                          0   

   Job_unskilled - resident  Telephone_none  Telephone_yes  foreign worker_no  \
0                         0               0              1                  0   
1                         0               1              0                  0   
2                         1               1              0                  0   
3                         0               1              0                  0   
4                         0               1              0                  0   

   foreign worker_yes  good_risk  
0                   1          1  
1                   1          0  
2                   1          1  
3                   1          1  
4                   1          0  

[5 rows x 63 columns]

In [5]:
#y_col = "good_risk"
y = df_new["good_risk"]

X = df_new[df_new.columns.drop("good_risk")]

In [6]:
from sklearn. model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2)

In [7]:
display(Markdown("#### Training Dataset shape"))
print(X_train.shape)
display(Markdown("#### Testing Dataset shape"))
print(X_test.shape)


#### Training Dataset shape

(800, 62)


#### Testing Dataset shape

(200, 62)


In [8]:
# Create training dataset by standardizing columns
scaler = StandardScaler()

# Fit and scale features in the training set
X_train = scaler.fit_transform(X_train)
#y_train = y_train.ravel()

# Scale test data
X_test = scaler.transform(X_test)
#y_test = y_test.labels

### Logistic Regression Model

In [9]:
# Run Logistic Regression
lrmod = LogisticRegression()
model = lrmod.fit(X_train, y_train)

In [10]:
y_train_pred = model.predict(X_train)
y_pred = model.predict(X_test)

In [11]:
from sklearn.metrics import confusion_matrix

display(Markdown("### Results for the training dataset\n"))

print('\n confusion matrix training set')
print(confusion_matrix(y_train, y_train_pred))
print('\n accuracy score:', accuracy_score(y_train, y_train_pred))
print('\n f1 score:', f1_score(y_train, y_train_pred))


display(Markdown("### Results for the test dataset\n"))
print('\n confusion matrix test set')
print(confusion_matrix(y_test, y_pred))
print('\n accuracy score:', accuracy_score(y_test, y_pred))
print('\n f1 score:', f1_score(y_test, y_pred))

### Results for the training dataset



 confusion matrix training set
[[124 116]
 [ 63 497]]

 accuracy score: 0.77625

 f1 score: 0.8473998294970162


### Results for the test dataset



 confusion matrix test set
[[ 28  32]
 [ 23 117]]

 accuracy score: 0.725

 f1 score: 0.8096885813148789


#### Uitleg metrics:
<div>&nbsp;</div>
<div>- Confusion matrix:</div>
<div>&nbsp; &nbsp; Relatie weergeven tussen positieve en negatieve voorspellingen en de werkelijkheid:</div>
<div>
<table style="border-collapse: collapse; width: 44.0421%; height: 79px;" border="1">
<tbody>
<tr>
<td style="width: 49.9021%;">True Positive</td>
<td style="width: 49.9021%;">False Positive</td>
</tr>
<tr>
<td style="width: 49.9021%;">False Negative</td>
<td style="width: 49.9021%;">True Negative</td>
</tr>
</tbody>
</table>
&nbsp; &nbsp;</div>
<div>- Accuracy score:</div>
<div>&nbsp; &nbsp; Hoe vaak voorspelt het model correct?</div>
<div>&nbsp;</div>
<div>- F1 score:</div>
<div>&nbsp; &nbsp; Gemiddelde van Precision en Recall:</div>
<div style="padding-left: 40px;">&nbsp; &nbsp; Precision: als het model positief voorspelt, hoe vaak is dit correct?</div>
<div style="padding-left: 40px;">&nbsp; &nbsp; Recall: Deel van de waarden dat juist is voorspeld.</div>
<p><br /><br /></p>
<div>&nbsp; &nbsp;</div>
<p><br /><br /></p>

### XGBoost Model

In [12]:
#from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV

estimator = XGBClassifier(base_score=0.7, eval_metric="error", seed=1994)

parameters = {'colsample_bytree': [0.1, 0.5, 1],
              'min_child_weight': [5, 15, 25],
              'learning_rate': [0.001, 0.05, 0.1, 0.2, 0.30],
              'n_estimators': [10, 50, 100],
              'max_depth':[2, 4, 6],
              'gamma': [0, 0.001, 0.005]              
             }


grid_search = GridSearchCV(estimator=estimator,
                           param_grid=parameters,
                           scoring = 'balanced_accuracy',
                           n_jobs = 1,
                           cv = 10)

In [15]:
tune_xgb_params = True

# tune_xgb_params can be set to false, so no grid search will be performed in a next run
#tune_xgb_params = False

if tune_xgb_params:
    start = time.time()
    grid_search.fit(X_train, y_train)
    end = time.time()
    
    xgbmod  = grid_search.best_estimator_
    minutes= (end - start) /60
    print( "It took %.2f minutes" % minutes)

It took 12.30 minutes


In [25]:
print(xgbmod)

XGBClassifier(base_score=0.7, booster='gbtree', callbacks=None,
              colsample_bylevel=1, colsample_bynode=1, colsample_bytree=1,
              early_stopping_rounds=None, enable_categorical=False,
              eval_metric='error', gamma=0, gpu_id=-1, grow_policy='depthwise',
              importance_type=None, interaction_constraints='',
              learning_rate=0.2, max_bin=256, max_cat_to_onehot=4,
              max_delta_step=0, max_depth=4, max_leaves=0, min_child_weight=5,
              missing=nan, monotone_constraints='()', n_estimators=100,
              n_jobs=0, num_parallel_tree=1, predictor='auto',
              random_state=1994, reg_alpha=0, reg_lambda=1, ...)


In [26]:
#copy paste the output from the previous cell, leave out 'None' values > will throw error
#xgbmod = XGBClassifier(base_score=0.7, booster='gbtree', colsample_bylevel=1,
              #colsample_bynode=1, colsample_bytree=0.5,
              #enable_categorical=False, eval_metric='error', gamma=0, gpu_id=-1,
              #importance_type=None, interaction_constraints='',
              #learning_rate=0.3, max_delta_step=0, max_depth=2,
              #min_child_weight=15, monotone_constraints='()',
              #n_estimators=50, n_jobs=6, num_parallel_tree=1, predictor='auto',
              #random_state=1994, reg_alpha=0, reg_lambda=1, scale_pos_weight=1,
              #seed=1994, subsample=1, tree_method='exact',
              #validate_parameters=1, verbosity=None)


xgbmod.fit(X_train, y_train)

XGBClassifier(base_score=0.7, booster='gbtree', callbacks=None,
              colsample_bylevel=1, colsample_bynode=1, colsample_bytree=1,
              early_stopping_rounds=None, enable_categorical=False,
              eval_metric='error', gamma=0, gpu_id=-1, grow_policy='depthwise',
              importance_type=None, interaction_constraints='',
              learning_rate=0.2, max_bin=256, max_cat_to_onehot=4,
              max_delta_step=0, max_depth=4, max_leaves=0, min_child_weight=5,
              missing=nan, monotone_constraints='()', n_estimators=100,
              n_jobs=0, num_parallel_tree=1, predictor='auto',
              random_state=1994, reg_alpha=0, reg_lambda=1, ...)

In [29]:
xgb_y_train_pred = xgbmod.predict(X_train)

In [30]:
xgb_y_test_pred = xgbmod.predict(X_test)

In [31]:
display(Markdown("### Results for the training dataset\n"))

print('\n confusion matrix training set')
print(confusion_matrix(y_train, xgb_y_train_pred))
print('\n accuracy score:', accuracy_score(y_train, xgb_y_train_pred))
print('\n f1 score:', f1_score(y_train, xgb_y_train_pred))


display(Markdown("### Results for the test dataset\n"))
print('\n confusion matrix test set')
print(confusion_matrix(y_test, xgb_y_test_pred))
print('\n accuracy score:', accuracy_score(y_test, xgb_y_test_pred))
print('\n f1 score:', f1_score(y_test, xgb_y_test_pred))

### Results for the training dataset



 confusion matrix training set
[[194  46]
 [ 12 548]]

 accuracy score: 0.9275

 f1 score: 0.949740034662045


### Results for the test dataset



 confusion matrix test set
[[ 31  29]
 [ 19 121]]

 accuracy score: 0.76

 f1 score: 0.8344827586206898


This results show that the XGB model is overfitting on the training dataset: the accuracy for the training set is 93% and the accuracy for the test set is 76%